qemu/tests/block_hotplug.py

```diff
@@ -88,153 +76,81 @@ def unplug_device(vm, get_disk_cmd, device):
 
     for iteration in range(repeat_times):
         device_list = []
-        controller_list = []
-        controller_device_dict = {}
         error_context.context("Hotplug block device (iteration %d)" % iteration,
                               logging.info)
 
-        sub_type = params.get("sub_type_before_plug")
-        if sub_type:
-            error_context.context(context_msg % (sub_type, "before hotplug"),
-                                  logging.info)
-            utils_test.run_virt_sub_test(test, params, env, sub_type)
+        plug_tag = "before_plug"
+        run_sub_test(params, plug_tag)
 
         for num in range(blk_num):
-            device = qdevices.QDevice(pci_type)
+            image_name = img_list[num + 1]
+            image_params = params.object_params(image_name)
+            drive_format = image_params.get("drive_format")
             if params.get("need_plug") == "yes":
                 disks_before_plug = find_disk(vm, get_disk_cmd)
-
-                if params.get("need_controller", "no") == "yes":
-                    controller_model = params.get("controller_model")
-                    controller = qdevices.QDevice(controller_model, params={"id":
-                                                  "hotadded_scsi%s" % num})
-                    bus_extra_param = params.get("bus_extra_params_%s" % img_list[num + 1])
-                    if bus_extra_param:
-                        for item in bus_extra_param.split():
-                            key, value = item.split("=", 1)
-                            qdevice_params = {key: value}
-                            controller.params.update(qdevice_params)
-                    controller.hotplug(vm.monitor)
-                    ver_out = controller.verify_hotplug("", vm.monitor)
-                    if not ver_out:
-                        err = "%s is not in qtree after hotplug" % controller_model
-                        test.fail(err)
+                devs = vm.devices.images_define_by_params(image_name,
+                                                          image_params, 'disk')
+                for dev in devs:
+                    ret = vm.devices.simple_hotplug(dev, vm.monitor)
+                    if ret[1] is False:
+                        test.fail("Unable to verify status of hotplugging '%s'."
```

Comment:
Well False means we verified the status and it failed, so this message is misleading. As mentioned on previous version, the perfect solution is:

```python
if False:
    test.fail("Fail to hotplug ...")
elif not True:
   logging.warning("Unable to verify ...")
```

but as we check later for the presence of the disks I'd simply go with `if ret[1] is False: test.fail("Failed to hotplug...`